hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+            self.log.warn('not a transaction', hash=tx.hash_hex)
+            # Not a transaction. Punish peer?
+            return
+
+        self._tx_received += 1
+        if self._tx_received > self._tx_max_quantity + 1:
+            self.log.warn('too many txs received')
+            self.state = PeerState.ERROR
+            return
+
+        try:
+            # this methods takes care of checking if the tx already exists, it will take care of doing at least
+            # a basic validation
+            # self.log.debug('add new tx', tx=tx.hash_hex)
+            if self.partial_vertex_exists(tx.hash):
+                # XXX: early terminate?
+                self.log.debug('tx early terminate?', tx_id=tx.hash.hex())
+            else:
+                self.log.debug('tx received', tx_id=tx.hash.hex())
+            self.on_new_tx(tx, propagate_to_peers=False, quiet=True, reject_locked_reward=True)
+        except HathorError:
+            self.log.warn('invalid new tx', exc_info=True)
+            # Invalid block?!
+            # Invalid transaction?!
+            # Maybe stop syncing and punish peer.
+            self.state = PeerState.ERROR
+            return
+        else:
+            # XXX: debugging log, maybe add timing info
+            if self._tx_received % 100 == 0:
+                self.log.debug('tx streaming in progress', txs_received=self._tx_received)
+
+    # -----------------------------------
+    # BEGIN: GET_DATA/DATA implementation
+    # -----------------------------------
+
+    @inlineCallbacks
+    def get_tx(self, tx_id: bytes) -> Generator[Deferred, Any, BaseTransaction]:
+        """Async method to get a transaction from the db/cache or to download it."""
+        tx = self._get_tx_cache.get(tx_id)
+        if tx is not None:
+            self.log.debug('tx in cache', tx=tx_id.hex())
+            return tx
+        try:
+            tx = self.tx_storage.get_transaction(tx_id)
+        except TransactionDoesNotExist:
+            tx = yield self.get_data(tx_id, 'mempool')
+            if tx is None:
+                self.log.error('failed to get tx', tx_id=tx_id.hex())
+                self.protocol.send_error_and_close_connection(f'DATA mempool {tx_id.hex()} not found')
+                raise
+            if tx.hash != tx_id:
+                self.protocol.send_error_and_close_connection(f'DATA mempool {tx_id.hex()} hash mismatch')
+                raise
+        return tx
+
+    def get_data(self, tx_id: bytes, origin: str) -> Deferred:
+        """Async method to request a tx by id"""
+        # TODO: deal with stale `get_data` calls
+        if origin != 'mempool':
+            raise ValueError(f'origin={origin} not supported, only origin=mempool is supported')
+        key = f'{origin}:{tx_id.hex()}'
+        deferred = self.deferred_by_key.get(key, None)
+        if deferred is None:
+            deferred = self.deferred_by_key[key] = Deferred()
+            self.send_get_data(tx_id, origin=origin)
```

Comment:
In case we lose connection to the peer while waiting for the response for this GET-DATA request, the deferred will not be removed from the `self.deferred_by_key` dict.

Is this something we have to worry about and should we clean them when the sync manager is stopped? Or will they be automatically garbage collected?

Reply:
You're right. We should clean all these deferred items when the connection is lost.